app/com/arpnetworking/rollups/ConsistencyChecker.java

```diff
@@ -0,0 +1,579 @@
+        final double fractionalDataLoss = nSamplesDropped / nOriginalSamples;
+        metrics.setGauge("rollup/consistency_checker/fractional_data_loss", fractionalDataLoss);
+        final LogBuilder logBuilder =
+                // TODO(spencerpearson): probably make this level-thresholding configurable?
+                fractionalDataLoss == 0 ? LOGGER.trace()
+                : fractionalDataLoss < 0.001 ? LOGGER.debug()
+                : fractionalDataLoss < 0.01 ? LOGGER.info()
+                : fractionalDataLoss < 0.1 ? LOGGER.warn()
+                : LOGGER.error();
+        logBuilder.setMessage((fractionalDataLoss == 0 ? "no " : "") + "data lost in rollup")
+                .addData("task", task)
+                .addData("nOriginalSamples", nOriginalSamples)
+                .addData("nSamplesDropped", nSamplesDropped)
+                .addData("fractionalDataLoss", fractionalDataLoss)
+                .log();
+
+        /* TODO(spencerpearson, OBS-1176): re-trigger re-execution of bad datapoints, something like
+
+            if (discrepancyTooBig) {
+                _rollupManagerPool.tell(
+                    new RollupDefinition.Builder()
+                            .setSourceMetricName(task.getSourceMetricName())
+                            .setDestinationMetricName(task.getRollupMetricName())
+                            .setStartTime(task.getStartTime())
+                            .setPeriod(task.getPeriod())
+                            .build()
+                );
+            }
+
+         */
+    }
+
+    /* package private */ static SampleCounts parseSampleCounts(
+            final Task task,
+            final MetricsQueryResponse response
+    ) throws MalformedSampleCountResponse {
+        final Map<String, Long> countsByMetric = Maps.newHashMap();
+        if (response.getQueries().size() != 2) {
+            throw new MalformedSampleCountResponse("expected exactly 1 query, got " + response.getQueries().size(), response);
+        }
+        for (final MetricsQueryResponse.Query query : response.getQueries()) {
+            if (query.getResults().size() != 1) {
+                throw new MalformedSampleCountResponse("expected exactly 1 result, got " + query.getResults().size(), response);
+            }
+            final MetricsQueryResponse.QueryResult queryResult = query.getResults().get(0);
+            if (queryResult.getValues().size() != 1) {
+                throw new MalformedSampleCountResponse("expected exactly 1 value, got " + queryResult.getValues().size(), response);
+            }
+            final Optional<Object> value = queryResult.getValues().get(0).getValue();
+            if (!value.isPresent()) {
+                throw new MalformedSampleCountResponse("sample count has null value", response);
+            }
+            final long longValue;
+            try {
+                longValue = Long.parseLong(value.get().toString());
```

Comment:
Is the value always a long? Will this parse scientific notation if it's a very large double not representable as a long?